### PR TITLE
test: Initial PR for integrating the Page Object Model (POM) into e2e test suite

### LINKED ISSUE #24985: [POM] First PR to create new Page Object Model standard
### What is this about?

First PR to create new Page Object Model standard, share it and discuss.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by chloeYue**
## **Description**

The Page Object Model (POM) is a design pattern that promotes code reusability and maintainability in e2e testing by encapsulating UI element selectors and interactions within page objects. This initial PR is the beginning of integrating POM into e2e test suite, aiming to enhance test code quality.

The entire implementation is done in TypeScript instead of JavaScript.

Key Considerations for Implementation:

- Easy Adaptation: ensure that the structure is straightforward and intuitive, thereby accelerating the development speed and facilitating easier migration.

- Enhanced Logging: offer better insights during test execution and debugging, making it easier to investigate flaky test

Structure:
The POM structure is organized around distinct page objects for each application UI screen. (Common components, such as the HeaderNavbar, are directly integrated as part of a screen's component when interaction is required. This approach eliminates the need for explicit class extension and allows for the exclusion of the HeaderNavbar in screens where interaction with it is unnecessary.)

Page functions and process:
I've introduced page objects, each designed to encapsulate the elements and interactions specific to their respective pages. Additionally, I've implemented processes such as `loginWaitBalance` and `sendTransaction` to efficiently manage common test flows that require interactions across multiple screens.

Processes should be defined for sequences that involve multiple page objects, facilitating broader testing objectives like completing transactions or logging in. These are typically actions that navigate through several screens.

Functions within a class (page object) are best used for actions and verifications that are specific to a single page. This approach promotes the encapsulation and reusability of code for individual UI components or screens, making the tests more modular and maintainable.

3 Migrated Test Cases:
Migrated 3 transaction test cases to the new POM structure, showcasing the improved test architecture and log information. These migrations serve as a POC, demonstrating the effectiveness of POM in our testing environment and provide the base for future migrations.

Additionally, I'm eliminating CSS selectors that aren't robust and replacing them with more stable selectors in this PR.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25373?quickstart=1)

## **Related issues**

Fixes: #24985 
Relates to: #22464 

## **Manual testing steps**

Tests should pass on CI. Code should be easy to understand.

### **Before**

Screenshot 2024-06-14 at 12 01 32

Screenshot 2024-06-18 at 11 25 05

### **After**

Screenshot 2024-06-14 at 22 38 22

Screenshot 2024-06-18 at 11 28 10

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/page-objects/pages/send-token-page.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    console.log(
+      `Fill recipient input with ${recipientAddress} on send token screen`,
+    );
+    await this.driver.pasteIntoField(this.inputRecipient, recipientAddress);
+  }
+
+  async fillAmount(amount: string): Promise<void> {
+    console.log(`Fill amount input with ${amount} on send token screen`);
+    await this.driver.waitForSelector(this.inputAmount);
+    await this.driver.pasteIntoField(this.inputAmount, amount);
```
Comment (DDDDDanica) [inline test/e2e/page-objects/pages/send-token-page.ts:60]: Suggestion: can we add validate the input after paste into the field before submitting? do you think this would be valuable?

**Comment by DDDDDanica** [comment]
🔥

File: test/e2e/page-objects/pages/send-token-page.ts (hunk 1 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/send-token-page.ts:61]: Great catch, i have added the check in the latest commit, please check: https://github.com/MetaMask/metamask-extension/pull/25373/commits/c27938a2cb54cf5abcad98aab8e1b414d6ca6c3f

File: test/e2e/page-objects/pages/send-token-page.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  async check_ensAddressAsRecipient(
+    ensDomain: string,
+    address: string,
+  ): Promise<void> {
+    // click to select the resolved adress
+    await this.driver.clickElement({
+      text: ensDomain,
+      css: this.ensResolvedName,
+    });
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply test/e2e/page-objects/pages/send-token-page.ts:114]: This selector is the resolved name for an ens domain, we combine it with the expected ens domain text to ensure we are clicking the expected ens address. It is robust in my opinion. This approach aligns with our second selector strategy outlined [here](https://github.com/MetaMask/contributor-docs/blob/main/docs/e2e/extension-e2e-guidelines.md#guidelines): check the expected text as it is part of the user's experience. I'm open to further discussion if you have a different perspective.

File: test/e2e/page-objects/pages/homepage.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    console.log(
+      `Expected balance ${expectedBalance} ETH is displayed on homepage`,
+    );
+  }
+
+  async getBalance(): Promise<number> {
+    const element = await this.driver.findElement(this.balance);
+    await this.driver.waitForNonEmptyElement(element);
+    return parseFloat(await element.getText());
+  }
```
Comment (HowardBraham) [inline test/e2e/page-objects/pages/homepage.ts:76]: This is a dangerous pattern that can cause race conditions and flakiness.  You want to search for the text you expect to find, not just get the current text and compare it.

File: test/e2e/page-objects/pages/homepage.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+   * @returns A promise that resolves if the expected number of confirmed transactions is displayed within the timeout period.
+   */
+  async check_confirmedTxNumberDisplayedInActivity(
+    expectedNumber: number = 1,
+  ): Promise<void> {
+    console.log(
+      `Wait for ${expectedNumber} confirmed transactions to be displayed in activity list`,
+    );
+    await this.driver.wait(async () => {
+      const confirmedTxes = await this.driver.findElements(
```
Comment (HowardBraham) [inline test/e2e/page-objects/pages/homepage.ts:101]: The short form of `Transactions` is normally `Txs`, not `Txes`

**Review by sleepytanya — DISMISSED** [review]

**Comment by hjetpoluru** [comment]
Thank you, @chloeYue for your effort. I have a small suggestion: could we adopt the model used by the mobile team? 

It could help developers working across both repositories by providing commonality, naming convention and making maintenance easier. cc @cortisiko 
Screenshot 2024-06-21 at 4 32 22 PM

File: test/e2e/constants.ts [hunk 5]
```diff
@@ -35,3 +35,6 @@ export const TEST_SNAPS_SIMPLE_KEYRING_WEBSITE_URL =
 
 /* Address of the VerifyingPaymaster smart contract deployed to Ganache. */
 export const VERIFYING_PAYMASTER = '0xbdbDEc38ed168331b1F7004cc9e5392A2272C1D7';
+
+/* Default ganache ETH balance in decimal when first login */
+export const DEFAULT_GANACHE_ETH_BALANCE_DEC = '25';
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply test/e2e/constants.ts:40]: Yes definitely! Changed, thanks!

File: test/e2e/page-objects/pages/login-page.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      css: '.unlock-page__title',
+      text: 'Welcome back!',
+    };
+  }
+
+  async check_pageIsLoaded(): Promise<void> {
+    try {
+      await this.driver.waitForSelector(this.welcomeBackMessage);
+      await this.driver.waitForSelector(this.passwordInput);
+      await this.driver.waitForSelector(this.unlockButton);
```
[the remark this answers is not in the document]
Comment (chloeYue) [reply test/e2e/page-objects/pages/login-page.ts:26]: I think it's a good suggestion ! I took your advice and wrote a new driver function `waitForMultipleSelectors()` to improve the performance. Please check.

File: test/e2e/page-objects/pages/homepage.ts (hunk 4 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/homepage.ts:101]: Changed, thanks !

File: test/e2e/page-objects/pages/homepage.ts (hunk 3 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/homepage.ts:76]: Nice catch! I agree it can cause flakiness. Initially, it was used in the function `check_expectedBalanceIsDisplayed()`. When the expected balance is not displayed, we call it to get the actual balance and log it in the error message. Your comment makes me realized that people can call it separately, which indeed will cause flakiness. I have removed this function, so we will force people to use `check_expectedBalanceIsDisplayed()`.

File: test/e2e/page-objects/pages/confirm-tx-page.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    this.confirmButton = '[data-testid="page-container-footer-next"]';
+    this.editButton = '[data-testid="confirm-page-back-edit-button"]';
+    this.transactionFee = '[data-testid="confirm-gas-display"]';
+  }
+
+  async check_pageIsLoaded(): Promise<void> {
+    try {
+      await this.driver.waitForSelector(this.editButton);
+      await this.driver.waitForSelector(this.confirmButton);
+      await this.driver.waitForSelector(this.transactionFee);
```
Comment (seaona) [inline test/e2e/page-objects/pages/confirm-tx-page.ts:23]: I am thinking if this could introduce a race condition, since having the element present in the DOM doesn't mean that ie its value is currently loaded, or the button is enabled etc.


I think for the buttons is fine, since we already have the guards for being buttons enabled in the click methods, however it could bring confusion with the transactionFee value or the total amount value if we add it.

Ie, in occasions we display a tx fee value, which is then updated after few seconds. So I wonder what makes a page 'loaded' :thinking: 
I think this depends on each screen, so we might need to make it slightly different case by case: in confirmation's page, we would consider it fully "loaded" when:
- we have the correct total amount 
- the correct fees expected
- the confirm button enabled (redundant with the click method, but ok to have)
So in this case, we could pass both values to this function from each test

`async check_pageIsLoaded(totalExpectedAmount, totalExpectedFees)`
and then look for the selector with the value passed.

This will be really good, since will mitigate all the race conditions related to this issue in the page level and ensure no more occurrences are introduced in the tests using this page

File: test/e2e/page-objects/processes/login.process.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import { WALLET_PASSWORD } from '../../helpers';
+
+/**
+ * This method unlocks the wallet and verifies that the user lands on the homepage with the expected balance. It is designed to be the initial step in setting up a test environment.
+ *
+ * @param driver - The webdriver instance.
+ * @param password - The password used to unlock the wallet. Defaults to WALLET_PASSWORD.
+ * @param expectedBalance - The expected balance to be displayed on the homepage after successful login. Defaults to DEFAULT_GANACHE_ETH_BALANCE_DEC, reflecting common usage in test setups.
+ */
+export const loginWaitBalance = async (
```
Comment (seaona) [inline test/e2e/page-objects/processes/login.process.ts:14]: nit: could this naming be `loginWithBalanceValidaiton` since it's essentially the same?

File: test/e2e/page-objects/processes/login.process.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  driver: Driver,
+  password: string = WALLET_PASSWORD,
+  expectedBalance: string = DEFAULT_GANACHE_ETH_BALANCE_DEC,
+) => {
+  console.log('Navigate to unlock page and try to login with pasword');
```
Comment (seaona) [inline test/e2e/page-objects/processes/login.process.ts:19]: could we initialize all the needed pages in a constructor? Something like:

```
constructor(driver) {
    this.driver = driver;
    this.loginPage = new LoginPage(driver);
    this.homePage = new HomePage(driver);
  }
  ```

**Comment by chloeYue** [comment]
> Thank you, @chloeYue for your effort. I have a small suggestion: could we adopt the model used by the mobile team?
> 
> It could help developers working across both repositories by providing commonality, naming convention and making maintenance easier. cc @cortisiko Screenshot 2024-06-21 at 4 32 22 PM

Hi @hjetpoluru, thanks for your suggestion! I've reviewed the page organization for Mobile and agree that alignment is beneficial. I've now created a `send` folder for all transaction-related pages, like the sendToken and confirmTransaction pages. Other pages, like the homepage, remain in the root for now. Regarding selectors, I prefer keeping them in the same file as the page functions. Since the selector strategy differs between extension and mobile, this approach simplifies development by reducing the need to switch files frequently to copy variable names.

File: test/e2e/page-objects/pages/confirm-tx-page.ts (hunk 7 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/confirm-tx-page.ts:23]: Yes, i agree confirmation page is a special case that need more check to mitigate all race condition issues. I adjusted as you suggested ! Thanks !

File: test/e2e/page-objects/processes/login.process.ts (hunk 8 above)
Comment (chloeYue) [reply test/e2e/page-objects/processes/login.process.ts:14]: Hi, i think the plan is to migrate as much code as possible to TypeScript. The new method is written in TS and includes a default balance value and enhanced logging. It is also designed in line with the POM pattern, it aims to simplify the adoption of POM, ensuring all new tests fully follow the POM pattern, including the processes they call. We will permit both methods to exist, and promoting the use of the new TS POM method, until full migration is accomplished.

File: test/e2e/page-objects/processes/login.process.ts (hunk 9 above)
Comment (seaona) [reply test/e2e/page-objects/processes/login.process.ts:19]: oh you are right! looks good as it is :see_no_evil:

**Review by DDDDDanica — DISMISSED** [review]

File: test/e2e/page-objects/pages/homepage.ts (hunk 4 above)
Comment (HowardBraham) [reply test/e2e/page-objects/pages/homepage.ts]: It looks like there were 4 `Txes`, but you only fixed 2

File: test/e2e/page-objects/pages/login-page.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (HowardBraham) [reply test/e2e/page-objects/pages/login-page.ts]: Your async/await logic in `waitForMultipleSelectors` is not right.  Consider this simple example, equivalent to what you wrote:
```
await Promise.all([
  await new Promise((resolve) => setTimeout(resolve, 1000)),
  await new Promise((resolve) => setTimeout(resolve, 1000)),
  await new Promise((resolve) => setTimeout(resolve, 1000)),
]);
```
That takes 3 seconds to execute.  You want it to do this, which takes 1 second to execute:
```
await Promise.all([
  new Promise((resolve) => setTimeout(resolve, 1000)),
  new Promise((resolve) => setTimeout(resolve, 1000)),
  new Promise((resolve) => setTimeout(resolve, 1000)),
]);
```
But also... is there an advantage to the way you wrote it (independent of `waitForSelector`) vs writing it like this:
```
async waitForMultipleSelectors(
  rawLocators,
  { timeout = this.timeout, state = 'visible' } = {},
) {
  const promises = rawLocators.map((rawLocator) =>
    this.waitForSelector(rawLocator, { timeout, state }),
  );

  return Promise.all(promises);
}
```

File: test/e2e/page-objects/processes/login.process.ts (hunk 8 above)
Comment (seaona) [reply test/e2e/page-objects/processes/login.process.ts:14]: oh yeah I'm okay with having both methods pre-existing. I was suggesting to make the name `loginWaitBalance` a bit more explicit like the other one, but it's also fine with me as it is :+1:

**Review by seaona — DISMISSED** [review]

**Comment by hjetpoluru** [comment]
@chloeYue, let me share a screenshot of another folder structure in the development code, which will help you understand the selectors I am mentioning, this is specifically the getters/setters. For easy maintenance we need to organize this folder structure, reduces the duplication of the selectors and readable.
Screenshot 2024-06-26 at 8 56 49 AM

File: test/e2e/page-objects/pages/homepage.ts (hunk 4 above)
Comment (chloeYue) [reply test/e2e/page-objects/pages/homepage.ts]: oops, yes i missed it, changed ! Thanks Howard!

File: test/e2e/page-objects/processes/login.process.ts (hunk 8 above)
Comment (chloeYue) [reply test/e2e/page-objects/processes/login.process.ts:14]: oh i guess i misunderstood you ! sure, makes sense, i changed the name to `loginWithBalanceValidaiton` ! Thanks!

File: test/e2e/page-objects/pages/login-page.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (chloeYue) [reply test/e2e/page-objects/pages/login-page.ts]: Hi @HowardBraham , Indeed, I should call `waitForSelector` directly in my new function! I have made the adjustments based on your comment. Could you please take another look to see if that's what you mean? Thanks !

**Comment by chloeYue** [comment]
> @chloeYue, let me share a screenshot of another folder structure in the development code, which will help you understand the selectors I am mentioning, this is specifically the getters/setters. For easy maintenance we need to organize this folder structure, reduces the duplication of the selectors and readable. Screenshot 2024-06-26 at 8 56 49 AM

Thank you for your suggestion. I understand the structure you're proposing. At this current phase, I believe it's beneficial to define the selectors within the same file as the page functions. 
- If you review the implementation in this PR, you'll notice that all interactions with selectors occur within the page functions themselves, we never need to reference a selector from other files, which is why we've encapsulated them with interaction functions in same file.
- As we're in the adoption phase of the POM, keeping the selectors and page functions in a single file significantly reduces the need to switch files just to copy selector variable names.
- Regarding the concerns about readable, maintenance and duplication you mentioned, I don't see these as issues with current organization. It's readable, as easy to maintain as the structure shown in your screenshot, and there's no duplication. We can always update selectors in the definition line, call `this.selector1` in page functions and call page functions in tests, writing all setting all getters seems redundant and not necessary in this case.
- The biggest challenge for adopting POM is ensuring it's user-friendly and easy for everyone to adapt to. We should aim for the most straightforward organization possible, keeping in mind not to over-engineer.
In conclusion, I believe it's better to keep the selectors organized in the current way during this phase.

File: test/e2e/page-objects/pages/login-page.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (HowardBraham) [reply test/e2e/page-objects/pages/login-page.ts]: Yes, all my comments have been addressed now, thanks!

**Comment by hjetpoluru** [comment]
Thanks @chloeYue for explaining. I agree to you this is also easy pattern to maintain as it available in the page and any flakiness would be resolved.

**Review by DDDDDanica — APPROVED** [review]

**Review by HowardBraham — APPROVED** [review]

**Review by seaona — APPROVED** [review]
Awesome job!
